syntax = "proto3";

package gateway;

message LearningParameter {
    int32 algorithm = 1;
    int32 feedback_algorithm = 2;
}

message Result {
    int32 question_answer_id = 1;
    float probability = 2;
    string question = 3;
}

message SetupRequest {}
message SetupResponse {}

message ReplyRequest {
    int32 bot_id = 1;
    string body = 2;
    LearningParameter learning_parameter = 3;
}

message ReplyRequests {
    repeated ReplyRequest data = 1;
}

message ReplyResponse {
    int32 question_feature_count = 1;
    repeated Result results = 2;
    int32 noun_count = 3;
    int32 verb_count = 4;
    int32 algorithm = 5;
    int32 feedback_algorithm = 6;
    ProcessingResponse processing = 7;
    LearnMetaResponse meta = 8;
}

message ReplyResponses {
    repeated ReplyResponse data = 1;
}

message ProcessingResponse {
    repeated string tokenized = 1;
}

message LearnRequest {
    int32 bot_id = 1;
    LearningParameter learning_parameter = 3;
}

message LearnResponse {
    float accuracy = 2;
    float precision = 3;
    float recall  = 4;
    float f1 = 5;
    LearnMetaResponse meta = 6;
}

message LearnMetaResponse {
    string env = 1;
    int32 bot_id = 2;
    int32 algorithm = 3;
    int32 feedback_algorithm = 4;
    LearnConfigResponse config = 5;
}

message LearnConfigResponse {
    bool word2vec_model_is_binary = 1;
    string dicdir = 2;
    string datasource_type = 3;
    string word2vec_model_name = 4;
}

service Bot {
    rpc Setup (SetupRequest) returns (SetupResponse) {}
    rpc Reply (ReplyRequest) returns (ReplyResponse) {}
    rpc Replies (ReplyRequests) returns (ReplyResponses) {}
    rpc Learn (LearnRequest) returns (LearnResponse) {}
}
